File: 04-2-One2NineAppWithRoom/app/src/main/java/com/example/one2nineapp/persistence/ScoreDao.java
package com.example.one2nineapp.persistence;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;

@Dao
public interface ScoreDao {

    // select all scores and observes for changes
    @Query("SELECT * FROM scores ORDER BY time ASC")
    public Flowable<List<Score>> getAllScores();

    // return when the insertion is completed
    @Insert
    public Completable insertScore(Score score);

    // return the number of rows affected
    @Query("DELETE FROM scores")
    public Single<Integer> deleteAllScores();

    // select all scores
    @Query("SELECT * FROM scores ORDER BY time ASC LIMIT 1")
    public Single<Score> getBestScore();

}

File: kotlin_projects/04-2-One2NineAppWithRoom_kt2/app/src/main/java/com/example/one2nineappwithroom_kt2/repository/ScoreDao.kt
package com.example.one2nineappwithroom_kt2.repository

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import kotlinx.coroutines.flow.Flow

@Dao
interface ScoreDao {
    @Insert
    suspend fun insert(score: Score)

    @Query("SELECT * FROM Score ORDER BY time ASC LIMIT 1")
    suspend fun getBestScore(): Score

    @Query("SELECT * FROM Score ORDER BY time ASC")
    fun getAllScores(): Flow<List<Score>>

    @Query("DELETE FROM Score")
    suspend fun deleteAllScores()
}